cleanlab/object_detection/filter.py

```diff
@@ -217,3 +234,146 @@ def _pool_box_scores_per_image(is_issue_per_box: List[np.ndarray]) -> np.ndarray
+        precisions = tp / np.maximum((tp + fp), eps)
+        recalls = recalls[0, :]
+        precisions = precisions[0, :]
+        ap = calculate_average_precision(recalls, precisions)
+        ap_per_class_list.append(ap)
+    if num_imgs > 1:
+        pool.close()
+    return ap_per_class_list
+
+
+def filter_by_class(labels: List[Dict[str, Any]], predictions: List[np.ndarray], class_num: int):
+    """
+    Filters predictions and labels based on a specific class number.
+    """
+    pred_bboxes = [img_res[class_num] for img_res in predictions]
+    lab_bboxes = []
+    for label in labels:
+        gt_inds = label["labels"] == class_num
+        lab_bboxes.append(label["bboxes"][gt_inds, :])
+    return pred_bboxes, lab_bboxes
+
+
+def _get_tp_fp(pred_bboxes: np.ndarray, lab_bboxes: np.ndarray, iou_threshold: float = 0.5):
+    """Calculates true positives (TP) and false positives (FP) for object detection tasks.
+    It takes predicted bounding boxes, ground truth bounding boxes, and an optional Intersection over Union (IoU) threshold as inputs.
+    """
+    num_dets = pred_bboxes.shape[0]
+    num_gts = lab_bboxes.shape[0]
+    num_scales = 1
+    tp = np.zeros((num_scales, num_dets), dtype=np.float32)
+    fp = np.zeros((num_scales, num_dets), dtype=np.float32)
+    if lab_bboxes.shape[0] == 0:
+        fp[...] = 1
+        return tp, fp
+    ious = _get_overlap_matrix(pred_bboxes, lab_bboxes)
+    ious_max = ious.max(axis=1)
+    ious_argmax = ious.argmax(axis=1)
+    sort_inds = np.argsort(-pred_bboxes[:, -1])
+    gt_covered = np.zeros(num_gts, dtype=bool)
+    for ind in sort_inds:
+        if ious_max[ind] >= iou_threshold:
+            matched_gt = ious_argmax[ind]
+            if not gt_covered[matched_gt]:
+                gt_covered[matched_gt] = True
+                tp[0, ind] = 1
+            else:
+                fp[0, ind] = 1
+        else:
+            fp[0, ind] = 1
+    return tp, fp
+
+
+def calculate_average_precision(recall_list: np.ndarray, precision_list: np.ndarray):
+    """Computes the average precision (AP) for a set of recall and precision values. It takes arrays of recall and precision values as inputs."""
+    recall_list = recall_list[np.newaxis, :]
+    precision_list = precision_list[np.newaxis, :]
+    num_scales = recall_list.shape[0]
+    ap = np.zeros(num_scales, dtype=np.float32)
+    zeros = np.zeros((num_scales, 1), dtype=recall_list.dtype)
+    ones = np.ones((num_scales, 1), dtype=recall_list.dtype)
+    mrec = np.hstack((zeros, recall_list, ones))
+    mpre = np.hstack((zeros, precision_list, zeros))
+    for i in range(mpre.shape[1] - 1, 0, -1):
+        mpre[:, i - 1] = np.maximum(mpre[:, i - 1], mpre[:, i])
+    for i in range(num_scales):
+        ind = np.where(mrec[i, 1:] != mrec[i, :-1])[0]
+        ap[i] = np.sum((mrec[i, ind + 1] - mrec[i, ind]) * mpre[i, ind + 1])
+    return ap
+
+
+def get_per_class_ap(labels: List[Dict[str, Any]], predictions: List[np.ndarray]):
```

Comment:
This function should be private `_get_per_class_ap()`

Reply:
https://github.com/cleanlab/cleanlab/blob/3024fad44f236c2f546610be35c3cfa5d011ce31/cleanlab/object_detection/filter.py#L370